halo2_proofs/src/plonk/prover.rs

```diff
@@ -55,7 +55,20 @@ pub fn create_proof<
     instances: &[&[&[Scheme::Scalar]]],
     mut rng: R,
     transcript: &mut T,
-) -> Result<(), Error> {
+) -> Result<(), Error>
+where
+    Scheme::Scalar: PrimeField,
+{
+    #[cfg(feature = "counter")]
+    {
+        use crate::{FFT_COUNTER, MSM_COUNTER};
+        use std::collections::BTreeMap;
+
+        // reset counters at the beginning of the prove
+        *MSM_COUNTER.lock().unwrap() = BTreeMap::new();
+        *FFT_COUNTER.lock().unwrap() = BTreeMap::new();
+    }
+
```

Comment:
Do I correctly understand that at the line 81 ("let instance: Vec<InstanceSingle<Scheme::Curve>> = instances")  the array describing the sets of the instance collumns for each circuit intsance is converted into the array of corresponding polynomials, each of which is given in two forms: coefficients of the powers of x and coefficients of the Lagrange basis polynomials? And what is the boolean variable "P::QUERY_INSTANCE"?

Reply:
I think so yes.

`P::QUERY_INSTANCE` seems related to the commitment scheme. It is `true` for IPA and `false` for KZG. Why or what it does however, I do not know because this also seems something new (or at least I can't remember).